src/threadval.cpp

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#if defined(HAVE_CONFIG_H)
+#include <config/bitcoin-config.h>
+#endif
+
+#include <threadval.h>
+
+/*  TODO: using thread_local changes the abi in ways that may not play nice
+    when the c++ stdlib is linked dynamically. Disable it until thorough
+    testing has been done. */
+#undef HAVE_THREAD_LOCAL
+
+#if defined(HAVE_THREAD_LOCAL)
+#include <atomic>
+#include <thread>
+
+#elif defined(HAVE_PTHREAD)
+#include <atomic>
+#include <pthread.h>
+
+#else
+#include <mutex>
+#include <thread>
+#include <unordered_map>
+#endif
+
+struct thread_data_type
+{
+    constexpr thread_data_type() = default;
+    thread_data_type(long id, std::string name) : m_id(id), m_name(std::move(name)){}
+
+    long m_id{0};
+    std::string m_name{"unnamed"};
+};
+
+#if defined(HAVE_THREAD_LOCAL)
```

Comment:
Do we want to use thread local even on mingw (where you had problems earlier?)

Reply:
https://github.com/bitcoin/bitcoin/pull/13168/files#diff-defae1c29982e5afaf477729b6c7d9b4R11 :)